datagen/src/main/scala/com/nvidia/rapids/tests/scaletest/TableGenerator.scala

```diff
@@ -0,0 +1,231 @@
+
+import scala.util.Random
+
+/**
+ * pre-concepts:
+ * - key groups:
+ * 1. The primary key for table a. In table a each row will be unique
+ * 2. Adjusted key group. The number of columns will correspond to the complexity. The types of the columns should be selected from the set {string, decimal(7, 2), decimal(15, 2), int, long, timestamp, date, struct<num: long, desc: string>}
+ * 3. 3-column key group with the types string, date, long.
+ * 4. 1 column that is an int with a unique count of 5
+ */
+class TableGenerator(scaleFactor: Int, complexity: Int, spark: SparkSession) {
+  val dbgen: DBGen = DBGen()
+  // define numRows in class as different table may depend on each other
+  private val aNumRows = scaleFactor * 10000
+  private val bNumRows = scaleFactor * 1000000
+  private val cNumRows = scaleFactor * 100000
+  private val dNumRows = scaleFactor * 100000
+  private val eNumRows = scaleFactor * 1000000
+  private val fNumRows = scaleFactor * 10000
+  private val gNumRows = scaleFactor * 1000000
+  private val seed = 41
+  private val random = new Random(seed)
+
+  private def randomColumnType(): String = {
+    // TODO: add more types when they are supported e.g. BinaryType CalendarintervalType etc.
+    // See: https://github.com/NVIDIA/spark-rapids/issues/8717
+    // ride-along/data columns will pick the column type from the candidates below.
+    // required at https://github.com/NVIDIA/spark-rapids/issues/8813#issue-1822958165
+    // TODO: maybe we can move this to an external text file so we can extend it easily in the future
+    val candidates = Seq(
+      "int",
+      "long",
+      "Decimal(7, 2)",
+      "Decimal(19, 4)",
+      "string",
+      "map<string, struct<lat: Decimal(7,4), lon: Decimal(7, 4)>>",
+      "timestamp",
+      "date")
+    candidates(random.nextInt(candidates.length))
+  }
+
+  private def randomNumericColumnType(): String = {
+    val candidates = Seq(
+      "int",
+      "long",
+      "Decimal(7, 2)",
+      "Decimal(19, 4)"
+    )
+    val seed = 41
+    val r = new Random(seed)
+    candidates(r.nextInt(candidates.length))
+  }
+
+  /**
+  a_facts: Each scale factor corresponds to 10,000 rows
+    - primary_a the primary key from key group 1
+    - key group 4
+    - a_data_low_unique_1 data column that is a long with a unique count of 5
+    - a_data_low_unique_len_1 data column that is a string with variable length ranging from 1 to 5 inclusive (even distribution of lengths)
+    - complexity data columns
+   */
+  private def genAFacts(): DataFrame = {
+    val schema = "primary_a long," +
+      "a_key4_1 int," +
+      "a_data_low_unique_1 long," +
+      "a_data_low_unique_len_1 string," +
+      (1 to complexity).map(i => s"a_data_$i ${randomColumnType()}").mkString(",")
+    val aFacts = dbgen.addTable("a_facts", schema, aNumRows)
+    aFacts("primary_a").setSeedRange(1, aNumRows)
```

Comment:
For this one I think we want to configure primary)_a with a DistinctDistribution.

```
 aFacts("primary_a").setSeedRange(1, aNumRows).setSeedMapping(DistinctDistribution())
```